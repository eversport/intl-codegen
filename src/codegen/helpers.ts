export function stable(json: unknown): string {
  if (typeof json === "object" && json) {
    const keys = Object.keys(json);
    keys.sort((a, b) => a.localeCompare(b));
    if (!keys.length) {
      return "{}";
    }
    return `{ ${keys
      .map(key => `${isId(key) ? key : JSON.stringify(key)}: ${stable((json as any)[key])}`)
      .join(", ")} }`;
  }
  return JSON.stringify(json);
}

export function camelify(str: string) {
  return str.replace(/[_-](\w|$)/g, (_, ch) => ch.toUpperCase());
}

const ID_RE = /^[A-Za-z_$][$\w]*$/;
export function isId(name: string) {
  return ID_RE.test(name);
}
