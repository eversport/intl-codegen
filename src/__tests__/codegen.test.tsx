import React from "react";
import { renderToStaticMarkup } from "react-dom/server";
import { ensureCompiledFixture, testTypings } from "./helpers";

// running typescript takes quite some time
jest.setTimeout(10 * 1000);

describe("Codegen", () => {
  ensureCompiledFixture("simple", async dir => {
    const { loadLanguage } = require(dir);
    let lang;

    lang = await loadLanguage("en");
    expect(lang.test()).toEqual("en");

    lang = await loadLanguage("de");
    expect(lang.test()).toEqual("de");

    lang = await loadLanguage("fallback");
    expect(lang.locale).toEqual("en");
    expect(lang.test()).toEqual("en");
  });

  ensureCompiledFixture("dashed-ids", async dir => {
    const { loadLanguage } = require(dir);
    let lang;

    lang = await loadLanguage("en");
    expect(lang.aDashedId()).toEqual("dashed!");
  });

  beforeEach(() => {
    jest.spyOn(console, "warn").mockImplementation(() => {});
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  ensureCompiledFixture("react", async dir => {
    const { loadLanguage, Consumer, Provider, Localized, withIntl } = require(dir);
    let lang, rendered;

    const params = { react: <strong>react</strong> };

    lang = await loadLanguage("en");
    rendered = renderToStaticMarkup(
      <Provider value={lang}>
        <Localized id="test" params={params} />
      </Provider>,
    );
    expect(rendered).toEqual("a <strong>react</strong> element");

    lang = await loadLanguage("de");
    rendered = renderToStaticMarkup(
      <Provider value={lang}>
        <Localized id="test" params={params} />
      </Provider>,
    );
    expect(rendered).toEqual("ein <strong>react</strong> element");

    lang = await loadLanguage("de");
    rendered = renderToStaticMarkup(
      <Provider value={lang}>
        <Consumer>
          {(intl: any) => {
            const text = intl.test({ react: "text" });
            return `${typeof text}: ${text}`;
          }}
        </Consumer>
      </Provider>,
    );
    expect(rendered).toEqual("string: ein text element");

    const MyComponent = withIntl(({ prop, intl }: { prop: string; intl: any }) => {
      const text = intl.test({ react: prop });
      return `${typeof text}: ${text}`;
    });

    lang = await loadLanguage("de");
    rendered = renderToStaticMarkup(
      <Provider value={lang}>
        <MyComponent prop="text" />
      </Provider>,
    );
    expect(rendered).toEqual("string: ein text element");

    rendered = renderToStaticMarkup(
      <>
        <Localized id="test" params={params} />
        <Localized id="test2" params={params} />
      </>,
    );
    expect(rendered).toEqual("");
    expect(console.warn).toHaveBeenCalledTimes(1);
    expect(console.warn).toHaveBeenCalledWith(
      "Localization not initialized correctly.\nMake sure to include `<Provider value={await loadLanguage(__LOCALE__)}>` in your component tree.",
    );

    jest.resetAllMocks();

    lang = await loadLanguage("de");
    rendered = renderToStaticMarkup(
      <Provider value={lang}>
        <>
          <Localized id="invalid" params={params} /> <Localized id="invalid" params={params} />{" "}
          <Localized id="invalid2" params={params} />
        </>
      </Provider>,
    );
    expect(rendered).toEqual("invalid invalid invalid2");
    expect(console.warn).toHaveBeenCalledTimes(2);
    expect(console.warn).toHaveBeenNthCalledWith(1, `The translation key "invalid" is not defined.`);
    expect(console.warn).toHaveBeenNthCalledWith(2, `The translation key "invalid2" is not defined.`);
  });

  testTypings("typings-correct");
  testTypings("typings-locales");
  testTypings("typings-no-strings");

  testTypings("typings-ids");
  testTypings("typings-withIntl");

  testTypings("typings-wrong-locale");
  testTypings("typings-wrong-id");
  testTypings("typings-wrong-param");
});
