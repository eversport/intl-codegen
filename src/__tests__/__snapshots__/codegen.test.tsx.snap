// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Codegen typings-correct 1`] = `Array []`;

exports[`Codegen typings-ids 1`] = `
Array [
  "typings-ids/index.tsx (7,3): Property 'invalid' does not exist on type 'MyType'.",
]
`;

exports[`Codegen typings-locales 1`] = `
Array [
  "typings-locales/index.tsx (8,10): This condition will always return 'false' since the types '\\"en\\"' and '\\"foo\\"' have no overlap.",
]
`;

exports[`Codegen typings-no-strings 1`] = `
Array [
  "typings-no-strings/index.tsx (9,8): Type '{}' is not assignable to type 'never'.",
]
`;

exports[`Codegen typings-withIntl 1`] = `
Array [
  "typings-withIntl/index.tsx (30,8): Type '{}' is not assignable to type 'Pick<MyProps, \\"prop\\">'.
  Property 'prop' is missing in type '{}'.",
  "typings-withIntl/index.tsx (31,8): Type '{}' is not assignable to type 'Pick<MyProps, \\"prop\\">'.
  Property 'prop' is missing in type '{}'.",
]
`;

exports[`Codegen typings-wrong-id 1`] = `
Array [
  "typings-wrong-id/index.tsx (7,8): Property 'nonExistant' does not exist on type 'Intl'.",
  "typings-wrong-id/index.tsx (11,18): Type '\\"non-existant\\"' is not assignable to type '\\"test\\"'.",
  "typings-wrong-id/index.tsx (13,23): Property 'nonExistant' does not exist on type 'Intl'.",
]
`;

exports[`Codegen typings-wrong-locale 1`] = `
Array [
  "typings-wrong-locale/index.tsx (3,14): Argument of type '\\"incorrect-locale\\"' is not assignable to parameter of type '\\"en\\"'.",
]
`;

exports[`Codegen typings-wrong-param 1`] = `
Array [
  "typings-wrong-param/index.tsx (7,15): Argument of type '{ prmtr: string; }' is not assignable to parameter of type '{ parameter: any; }'.
  Object literal may only specify known properties, and 'prmtr' does not exist in type '{ parameter: any; }'.",
  "typings-wrong-param/index.tsx (8,21): Argument of type '{ parameter: string; }' is not assignable to parameter of type '{ second: any; }'.
  Object literal may only specify known properties, and 'parameter' does not exist in type '{ second: any; }'.",
  "typings-wrong-param/index.tsx (9,25): Argument of type '{ parameter: string; }' is not assignable to parameter of type '{ second: any; }'.
  Object literal may only specify known properties, and 'parameter' does not exist in type '{ second: any; }'.",
  "typings-wrong-param/index.tsx (13,38): Type '{ prmtr: string; }' is not assignable to type '{ parameter: any; } | { second: any; }'.
  Object literal may only specify known properties, and 'prmtr' does not exist in type '{ parameter: any; } | { second: any; }'.",
  "typings-wrong-param/index.tsx (14,45): Type '{ id: \\"test-dashed\\"; params: { parameter: string; }; }' is not assignable to type '(IntrinsicAttributes & { id: \\"test\\"; params: { parameter: any; }; } & { children?: ReactNode; }) | (IntrinsicAttributes & { id: \\"test-dashed\\" | \\"testDashed\\"; params: { second: any; }; } & { ...; })'.
  Type '{ id: \\"test-dashed\\"; params: { parameter: string; }; }' is not assignable to type '{ id: \\"test-dashed\\" | \\"testDashed\\"; params: { second: any; }; }'.
    Types of property 'params' are incompatible.
      Type '{ parameter: string; }' is not assignable to type '{ second: any; }'.
        Object literal may only specify known properties, and 'parameter' does not exist in type '{ second: any; }'.",
  "typings-wrong-param/index.tsx (15,44): Type '{ id: \\"testDashed\\"; params: { parameter: string; }; }' is not assignable to type '(IntrinsicAttributes & { id: \\"test\\"; params: { parameter: any; }; } & { children?: ReactNode; }) | (IntrinsicAttributes & { id: \\"test-dashed\\" | \\"testDashed\\"; params: { second: any; }; } & { ...; })'.
  Type '{ id: \\"testDashed\\"; params: { parameter: string; }; }' is not assignable to type '{ id: \\"test-dashed\\" | \\"testDashed\\"; params: { second: any; }; }'.
    Types of property 'params' are incompatible.
      Type '{ parameter: string; }' is not assignable to type '{ second: any; }'.
        Object literal may only specify known properties, and 'parameter' does not exist in type '{ second: any; }'.",
  "typings-wrong-param/index.tsx (17,30): Argument of type '{ prmtr: string; }' is not assignable to parameter of type '{ parameter: any; }'.
  Object literal may only specify known properties, and 'prmtr' does not exist in type '{ parameter: any; }'.",
]
`;
